Add Deck.gl Layers and Utility Support for vue-maplibre-gL

MostafaGamalSayed opened the pull request:
#### **Summary of Changes**
This pull request enhances the package with the following additions:
1. **New Components for Deck.gl Layers**:
    - Added `MglDeckGeojsonLayer` component to integrate Deck.gl's `GeoJsonLayer` ([@commit:43bab18a]()).
    - Added `MglDeckArcLayer` component to integrate Deck.gl's `ArcLayer` ([@commit:2dba67d3]()).

2. **Core Composable for Layer Management**:
    - Introduced `useDeckLayer` composable to handle lifecycle and reactivity for Deck.gl layers within Maplibre ([@commit:462eb07f]()).

3. **Utility and Prop Configuration**:
    - Added `deck.layer.lib.ts`, which defines common and layer-specific props with utilities for Deck.gl layers, ensuring reusable, type-safe layer configurations ([@commit:9503c64c]()).

4. **New Dependencies**:
    - Added Deck.gl-related packages (`@deck.gl/core`, `@deck.gl/layers`, `@deck.gl/mapbox`) to the project (`package.json` updates) ([@commit:b0b75cca]()).

#### **Features**
- **New Deck.gl Layers**: Support for `ArcLayer` and `GeoJsonLayer`, integrated as reusable Vue components.
- **Dynamic Tooltip Support**: Enables reactive tooltips for layers using `getTooltip` functionality.
- **Interactive Map Layers**: Allows developers to add, update, and remove custom Deck.gl layers easily.
- **Composable Architecture**: Simplifies usage of Deck.gl layers with Maplibre in Vue projects.

#### **Motivation**
The goal of these changes is to expand the functionality of the `@indoorequal/vue-maplibre-gl` package by introducing seamless integration with Deck.gl layers. These enhancements bring extended interactivity, 3D visualization, and customization options for applications using the library. The composable and utility-driven design ensures easier implementation and extension of Deck.gl layers within Vue projects.
#### **Future Plans**
This is just the beginning of integrating Deck.gl into the `@indoorequal/vue-maplibre-gl` package. Over time, the implementation will be extended to include **support for all available Deck.gl layers**, ensuring the package becomes a comprehensive solution for Deck.gl and Maplibre users. Future layers to be added include `ScatterplotLayer`, `IconLayer`, `BitmapLayer`, and many more.
#### **Testing**
- The changes have been tested with dynamic datasets for both `ArcLayer` and `GeoJsonLayer`.
- Verified layer interactivity, tooltip rendering, and reactivity under different map states

francois2metz commented:
Hi,

Thanks for your contribution. This add interesting new features. I don't know if it should be integrated to the default package, as it make the library bigger. This is something to look at.
Could you rebase your branch and make sure that the command `yarn build`? I was not able to build it locally. 
Also one or two examples would be nice.

MostafaGamalSayed commented:
Hi,

Thanks for the feedback! I’ve addressed the requested changes:
- **Rebased the branch** against the latest `main` branch of the repo.
- Verified that the build works correctly using `yarn build` (fixed any issues encountered during the process).
- Added examples for the new layers:
    - Example for `MglDeckArcLayer` demonstrating its usage with dynamic data and tooltips.
    - Example for `MglDeckGeojsonLayer` showcasing its integration with Maplibre.

- Updated documentation to link to the examples.

Let me know if there’s anything else you’d like me to adjust!

francois2metz commented on yarn.lock:
Why did you remove the lockfile?

MostafaGamalSayed replied:
Sorry, I accidentally removed it.

francois2metz commented:
Hi,

Could rebase your branch against the master branch and not create merge commits? This way we'll clean the history and enable the build check.

MostafaGamalSayed commented:
Hi,

I've rebased my branch against the latest `master` and force-pushed the changes as requested. The commit history is now clean, with no merge commits.

Please let me know if there is anything else you'd like me to do. Thank you for your feedback and guidance!

francois2metz commented:
Hi,

Thanks for the rebase!

I tried to run the examples. You may want to add the magic comments at the top of the file to have a title and the description of the example (look at the others ones).

Not sure one, but it seems that the deck.gl layers are not correctly binded with maplibre zoom and viewport. Look at the video

https://github.com/user-attachments/assets/890c95b9-8e9e-464d-b277-55ff83955c33

MostafaGamalSayed commented:
Hi,
Thank you for the feedback! I’ve made some updates based on your suggestions:
- **Magic Comments**: I’ve added the magic comments (title and description) to the top of the example files to make them consistent with the others in the project.
- **ScatterplotLayer Example**: After reviewing the issue, I found that the previous example didn’t handle the Deck.gl bindings with Maplibre’s zoom and viewport correctly. I’ve replaced it with a different example that resolves this issue and works as expected.
- **Other Layers**: For the ArcLayer and GeoJsonLayer examples, I’ve tested them again, and they seem to work properly with Maplibre's zoom and viewport. If you’re still seeing issues, please let me know, and I can investigate further.

I’ve pushed the changes—please let me know if there’s anything else I should adjust!
Thank you for all your help and guidance.

francois2metz commented:
Hi, the examples are still working for me. Did you forget to push a change?

MostafaGamalSayed commented:
Hi,

To clarify, I only made changes to the **ScatterplotLayer** example in the file `deck-scatteredplot-layer.vue` since it was the only example that had issues with Deck.gl bindings to Maplibre (zoom and viewport). The problem in this example has been resolved, and I’ve replaced it with a working version.
As for the other examples—`deck-geojson-layer.vue` and `deck-arc-layer.vue`—they were already working correctly without any issues, so I didn’t make any changes to them. I’ve tested these examples again, and they produce the same output as shown in the official Deck.gl documentation.

If there’s a specific issue you’re encountering with any of the examples, please let me know, and I’ll be happy to investigate further!

francois2metz commented:
Hi,

I have the same zoom problem with the scatterplot and the 2 other examples. 

https://github.com/user-attachments/assets/80f192fb-81ed-44e2-80d5-04850d4f2b76

MostafaGamalSayed commented:
Hi,

I’ve worked on the zoom issue and fixed it now in all three examples (ScatterplotLayer, GeojsonLayer, and ArcLayer) in commit `79baaa57`.

The issue was caused by missing styles for **MapLibre GL**, which led to improper rendering and interaction with the map at certain zoom levels. I’ve now imported the relevant MapLibre GL default styles in all the examples, which resolves the problem. Additionally, I adjusted the map’s height to make the examples more consistent and visually clear.

Please let me know if you encounter any further issues with the examples!

francois2metz commented:
Hi,

Much better indeed!

francois2metz commented on lib/components/deck-layers/arc.layer.ts:
```diff
@@ -0,0 +1,31 @@
+/**
+ * Deck.gl Arc Layer
+ *
+ * See the [https://deck.gl/docs/api-reference/layers/arc-layer]
```
Could you format the link a markdown?

MostafaGamalSayed replied:
Done

francois2metz commented on lib/components/deck-layers/geojson.layer.ts:
```diff
@@ -0,0 +1,32 @@
+/**
+ * Deck.gl Geojson Layer
+ *
+ * See the [https://deck.gl/docs/api-reference/layers/geojson-layer]
```
Could you format the link a markdown?

MostafaGamalSayed replied:
Done

francois2metz commented on lib/components/deck-layers/scatteredPlot.layer.ts:
```diff
@@ -0,0 +1,33 @@
+/**
+ * Deck.gl ScatteredPlot Layer
+ *
+ * See the [https://deck.gl/docs/api-reference/layers/scatterplot-layer]
```
Could you format the link a markdown?

MostafaGamalSayed replied:
Done

francois2metz commented on lib/lib/deck.layer.lib.ts:
```diff
@@ -0,0 +1,786 @@
+  type Accessor,
+  type Material,
+} from "@deck.gl/core";
+import type { MjolnirEvent } from "mjolnir.js";
```
It shouldn't a dependency? Or is it possible to use another type?

MostafaGamalSayed replied:
**mjolnir.js** is an indirect dependency, it is installed as part of the dependency tree of **Deck.gl**.
[see this](https://github.com/visgl/deck.gl/blob/master/docs/upgrade-guide.md#user-input-handling)

MostafaGamalSayed commented on lib/components/deck-layers/line.layer.ts:
```diff
@@ -0,0 +1,31 @@
+/**
+ * Deck.gl Line Layer
+ *
+ * See the [https://deck.gl/docs/api-reference/layers/line-layer]
```
Done

MostafaGamalSayed commented:
Hi @francois2metz ,

I just wanted to follow up on my pull request as I haven’t received any feedback in a while. Please let me know if there are any additional changes needed from my side or if there are concerns about merging the PR.

Thank you for your time.
